----
# **<span style="color:DarkSlateBlue">PROYECTO BOOKING_REVIEWS</span>**
# **<span style="color:DarkSlateBlue">Análisis preliminar</span>**
----

---
---
## <span style="color:gray">**1. Importación de librerías**</span> 📂

In [1]:
# Tratamiento de datos
import numpy as np
import pandas as pd 

# Configuración de ruta
import sys
sys.path.append('../')

# Importación de funciones personalizadas
from src.soporte1_preliminar import eda_preliminar

---
---
## <span style="color:gray">**2. Carga de datos**</span> 📥

In [2]:
booking_raw = pd.read_csv("../data/raw/hotel_booking.csv")

reviews_raw = pd.read_csv("../data/raw/hotel_reviews.csv")


In [3]:
# Configuración para mostrar todas las columnas
pd.set_option('display.max_columns', None)

---
---


----
# <span style="color:DarkSlateBlue">**Desarrollo del proyecto - 1**</span>
----

---
---
## <span style="color:gray">**Exploración inicial de los datasets**</span> 🔎

Antes de comenzar el análisis exploratorio de datos (EDA), es importante realizar una inspección general de los datasets. Este primer contacto con los datos nos permite entender su estructura, el tipo de variables que contienen y detectar posibles problemas como valores faltantes o formatos incorrectos.

Para ello utilizamos varias funciones de Pandas que nos ayudan a obtener una visión rápida y completa del conjunto de datos.

---

## `booking_raw`

In [4]:
# Llamamos a la función eda_preliminar guardada en el archivo de sorporte
eda_preliminar(booking_raw)



👀 Visualización de las primeras y últimas filas



,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,name,email,phone-number,credit_card
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,3,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01,Ernest Barnes,Ernest.Barnes31@outlook.com,669-792-1661,************4322
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,4,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01,Andrea Baker,Andrea_Baker94@aol.com,858-637-6955,************9157
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,0.0,0,BB,GBR,Direct,Direct,0,0,0,A,C,0,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02,Rebecca Parker,Rebecca_Parker@comcast.net,652-885-2745,************3734


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,name,email,phone-number,credit_card
119387,City Hotel,0,34,2017,August,35,31,2,5,2,0.0,0,BB,DEU,Online TA,TA/TO,0,0,0,D,D,0,No Deposit,9.0,NaN,0,Transient,157.71,0,4,Check-Out,2017-09-07,Mary Morales,Mary_Morales@hotmail.com,395-518-4100,************1821
119388,City Hotel,0,109,2017,August,35,31,2,5,2,0.0,0,BB,GBR,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,89.0,NaN,0,Transient,104.40,0,0,Check-Out,2017-09-07,Caroline Conley MD,MD_Caroline@comcast.net,531-528-1017,************7860
119389,City Hotel,0,205,2017,August,35,29,2,7,2,0.0,0,HB,DEU,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,9.0,NaN,0,Transient,151.20,0,2,Check-Out,2017-09-07,Ariana Michael,Ariana_M@xfinity.com,422-804-6403,************4482




------------------------------------------------------------------------


🎲 Visualización de filas aleatorias



,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,name,email,phone-number,credit_card
6430,Resort Hotel,1,155,2016,June,24,10,2,5,2,0.0,0,BB,PRT,Online TA,TA/TO,0,0,0,E,E,0,No Deposit,240.0,NaN,0,Transient,99.00,0,1,Canceled,2016-03-30,Debra Chavez,Debra_Chavez@comcast.net,498-511-1171,************1831
99188,City Hotel,0,338,2016,October,41,7,1,2,2,0.0,0,BB,FRA,Offline TA/TO,TA/TO,0,0,0,A,A,0,No Deposit,12.0,NaN,68,Transient-Party,90.67,0,1,Check-Out,2016-10-10,Kevin Lewis,KevinLewis@zoho.com,628-792-1813,************9092
56341,City Hotel,1,240,2016,September,36,3,2,2,2,0.0,0,BB,NLD,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,9.0,NaN,0,Transient,95.63,0,1,Canceled,2016-08-24,Kimberly Scott,Kimberly_Scott21@yahoo.com,721-827-4926,************4195




------------------------------------------------------------------------


📏 Dimensión del DataFrame

Número de filas del DataFrame: 119390
Número de columnas del DataFrame: 36


------------------------------------------------------------------------


🧾 Información general

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'name', 'email',
       'phone-number', 'credit_card'],
      dtype='str')



------------------------------------------------------------------------


🔢 Estadísticas descriptivas de las variables numéricas



,count,mean,std,min,25%,50%,75%,max
is_canceled,119390.0,0.370,0.483,0.00,0.00,0.000,1.0,1.0
lead_time,119390.0,104.011,106.863,0.00,18.00,69.000,160.0,737.0
arrival_date_year,119390.0,2016.157,0.707,2015.00,2016.00,2016.000,2017.0,2017.0
arrival_date_week_number,119390.0,27.165,13.605,1.00,16.00,28.000,38.0,53.0
arrival_date_day_of_month,119390.0,15.798,8.781,1.00,8.00,16.000,23.0,31.0
stays_in_weekend_nights,119390.0,0.928,0.999,0.00,0.00,1.000,2.0,19.0
stays_in_week_nights,119390.0,2.500,1.908,0.00,1.00,2.000,3.0,50.0
adults,119390.0,1.856,0.579,0.00,2.00,2.000,2.0,55.0
children,119386.0,0.104,0.399,0.00,0.00,0.000,0.0,10.0
babies,119390.0,0.008,0.097,0.00,0.00,0.000,0.0,10.0




------------------------------------------------------------------------


🔡 Estadísticas descriptivas de las variables categóricas



,count,unique,top,freq
hotel,119390,2,City Hotel,79330
arrival_date_month,119390,12,August,13877
meal,119390,5,BB,92310
country,118902,177,PRT,48590
market_segment,119390,8,Online TA,56477
distribution_channel,119390,5,TA/TO,97870
reserved_room_type,119390,10,A,85994
assigned_room_type,119390,12,A,74053
deposit_type,119390,3,No Deposit,104641
customer_type,119390,4,Transient,89613


### **Informe de EDA preliminar – Dataset `booking_raw`**

**1. Información general del Dataset**

- El Dataset contiene 119390 registros y +30 variables

- Incluye columnas numéricas, categóricas y temporales.

- Se identifican valores nulos en `agent`, `company`, `country` y `children`.

- No presenta registros duplicados → dataset consistente.


**2. Hallazgos por tipo de variable**

Variables numéricas:

- `adr`: alta variabilidad (hasta 5400), indicando posibles outliers.
- `lead_time`: media de 104 días, reflejando reservas con bastante anticipación.
- `is_canceled`: media del 37% → alta tasa de cancelación.

Variables categóricas:

- `adults`, `children`, `babies`: predominan reservas de 1–2 adultos sin niños.
- `previous_cancellations`: valores bajos → pocos clientes reincidentes cancelando.
- `reserved_room_type` y `assigned_room_type` → las mantenemos para analizar posibles discrepancias entre la habitación reservada y la asignada, que pueden correlacionar con reviews negativas.
- Predominan clientes “transient” y reservas online.


**3. Transformaciones clave**

Estandarización:
- Variables categóricas → minúsculas + snake_case.
- `phone-number` → `phone_number`.

Eliminación:

- `arrival_date_week_number` y `arrival_date_day_of_month`

    Solo necesitamos el mes y año de llegada para poder unir este dataset con el de reviews. Por lo tanto, estas columnas se pueden eliminar.

- `agent` y `company`

    Representan los ID de las agencias y compañías. No aportan información relevante para nuestro análisis, por lo que se eliminarán.

- `days_in_waiting_list`

    La mayoría de reservas no pasan por lista de espera; además, esta variable describe un proceso interno del hotel, no tanto al cliente. Por lo que la variable aporta poca información.

- `required_car_parking_spaces` y `total_of_special_requests`

    El valor promedio es muy bajo, la desviación estándar es baja y los percentiles muestran que la mayoría de los huéspedes no requiere estas opciones. Estas columnas aportan poca información útil y se eliminan.

- `name`, `email`, `phone-number`, `credit_card` 

    Nos dan información sensible del huésped que no es que no aporta valor al análisis. 


Tipos de datos:
- `is_canceled`, `is_repeated_guest` → bool.
- `children` → int.
- `reservation_status_date` → datetime.      

Feature Engineering:

- `total_nights` → Número de noches totales. 
- `total_guests` → Total de personas por reserva. 


💡 **Conclusión:**

El dataset está bien estructurado y, tras limpieza y transformación, queda listo para análisis exploratorio avanzado.

---
## `reviews_raw`

In [5]:
# Llamamos a la función eda_preliminar guardada en el archivo de sorporte
eda_preliminar(reviews_raw)



👀 Visualización de las primeras y últimas filas



,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Russia,I am so angry that i made this post available...,397,1403,Only the park outside of the hotel was beauti...,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Ireland,No Negative,0,1403,No real complaints the hotel was great great ...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,Australia,Rooms are nice but for elderly a bit difficul...,42,1403,Location was good and staff were ok It is cut...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968


,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
515735,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/19/2015,8.1,Atlantis Hotel Vienna,Egypt,The ac was useless It was a hot week in vienn...,19,2823,No Positive,0,3,2.5,"[' Leisure trip ', ' Family with older childre...",715 day,48.203745,16.335677
515736,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/17/2015,8.1,Atlantis Hotel Vienna,Mexico,No Negative,0,2823,The rooms are enormous and really comfortable...,25,3,8.8,"[' Leisure trip ', ' Group ', ' Standard Tripl...",717 day,48.203745,16.335677
515737,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/9/2015,8.1,Atlantis Hotel Vienna,Hungary,I was in 3rd floor It didn t work Free Wife,13,2823,staff was very kind,6,1,8.3,"[' Leisure trip ', ' Family with young childre...",725 day,48.203745,16.335677




------------------------------------------------------------------------


🎲 Visualización de filas aleatorias



,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
261016,92 Rue La Boetie 8th arr 75008 Paris France,88,10/26/2016,8.5,Rochester Champs Elysees,Egypt,No Negative,0,570,great location lovely rooms amazing staff hig...,9,2,10.0,"[' Leisure trip ', ' Group ', ' Classic Room '...",281 day,48.872495,2.309155
312103,Delflandlaan 15 Slotervaart 1062 EA Amsterdam ...,947,3/9/2017,8.7,Best Western Premier Hotel Couture,United Kingdom,Its needs a bar,5,8177,Beds were great hotel was very clean,9,2,9.2,"[' Leisure trip ', ' Couple ', ' Standard Doub...",147 day,52.351114,4.841163
155702,34 Norfolk Place Paddington Westminster Boroug...,634,7/19/2017,7.9,Norfolk Towers Paddington,Australia,There was only one lift which was frustrating...,12,4065,We stayed for three nights in a Club Double R...,104,109,9.2,"[' Leisure trip ', ' Couple ', ' Club Double R...",15 days,51.516288,-0.172108




------------------------------------------------------------------------


📏 Dimensión del DataFrame

Número de filas del DataFrame: 515738
Número de columnas del DataFrame: 17


------------------------------------------------------------------------


🧾 Información general

<class 'pandas.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  str    
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  str    
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  str    
 5   Reviewer_Nationality                        515738 non-null  str    
 6   Negative_Review         

Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Review_Date',
       'Average_Score', 'Hotel_Name', 'Reviewer_Nationality',
       'Negative_Review', 'Review_Total_Negative_Word_Counts',
       'Total_Number_of_Reviews', 'Positive_Review',
       'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags',
       'days_since_review', 'lat', 'lng'],
      dtype='str')



------------------------------------------------------------------------


🔢 Estadísticas descriptivas de las variables numéricas



,count,mean,std,min,25%,50%,75%,max
Additional_Number_of_Scoring,515738.0,498.082,500.538,1.000,169.000,341.000,660.000,2682.000
Average_Score,515738.0,8.397,0.548,5.200,8.100,8.400,8.800,9.800
Review_Total_Negative_Word_Counts,515738.0,18.539,29.691,0.000,2.000,9.000,23.000,408.000
Total_Number_of_Reviews,515738.0,2743.744,2317.465,43.000,1161.000,2134.000,3613.000,16670.000
Review_Total_Positive_Word_Counts,515738.0,17.776,21.804,0.000,5.000,11.000,22.000,395.000
Total_Number_of_Reviews_Reviewer_Has_Given,515738.0,7.166,11.040,1.000,1.000,3.000,8.000,355.000
Reviewer_Score,515738.0,8.395,1.638,2.500,7.500,8.800,9.600,10.000
lat,512470.0,49.442,3.466,41.328,48.215,51.500,51.516,52.400
lng,512470.0,2.824,4.579,-0.370,-0.143,0.011,4.834,16.429




------------------------------------------------------------------------


🔡 Estadísticas descriptivas de las variables categóricas



,count,unique,top,freq
Hotel_Address,515738,1493,163 Marsh Wall Docklands Tower Hamlets London ...,4789
Review_Date,515738,731,8/2/2017,2585
Hotel_Name,515738,1492,Britannia International Hotel Canary Wharf,4789
Reviewer_Nationality,515738,227,United Kingdom,245246
Negative_Review,515738,330011,No Negative,127890
Positive_Review,515738,412601,No Positive,35946
Tags,515738,55242,"[' Leisure trip ', ' Couple ', ' Double Room '...",5101
days_since_review,515738,731,1 days,2585


### **Informe de EDA preliminar – Dataset `reviews_raw`**

**1. Información general del Dataset**

- El dataset contiene 515738 registros y 17 variables.

- Presenta un bajo porcentaje de valores nulos en `lat` y `lng` (~0.63%).

- Existen algunos duplicados, pero la estructura general es consistente.

**2. Hallazgos por tipo de variable**

Variables numéricas:

- `Reviewer_Score` y `Average_Score`: media ≈ 8.4 → alto nivel de satisfacción general.
- `Total_Number_of_Reviews`: gran dispersión → diferentes niveles de popularidad de los hoteles.

Variables de texto y categóricas:

- `Positive_Review` y `Negative_Review`: alta variabilidad en contenido.
- `Reviewer_Nationality`: predominan usuarios del Reino Unido, posible sesgo geográfico.

**3. Transformaciones clave**

Estandarización:

- Nombres de las columnas → minúsculas.
- Normalización de `reviewer_nationality`: snake_case para homogeneizar valores (por ejemplo, *united_kingdom*).

Eliminación:

- `review_total_negative_word_counts` y `review_total_positive_word_counts`: no aportan valor para nuestro análisis.
- `days_since_review`: variable dependiente del momento de extracción, sin valor analítico.
- `lat` y `lng`: redundancia, ya tenemos la localización.

Tipos de datos:

- `review_date` → datetime.

Feature engineering:

- `city` y `country` → a partir de la dirección del hotel.
- Simplificación de `positive_review` y `negative_review` a palabras clave.
- `trip_type`, `group_type`, `room_type`, `nights` y `device` → a partir de la columna `tags`. 

💡 Conclusión:

El dataset es rico en información y adecuado para análisis de sentimiento. Tras limpieza, estandarización y creación de nuevas variables, queda optimizado para análisis avanzado e integración con bookings.